enhancements/sig-architecture/136-placement-cel-selector/README.md

```diff
@@ -0,0 +1,235 @@
+		// The input are 2 versions with type `cel.StringType`.
+		// The output is bool.
+		cel.Function("versionIsLessThan",
+			cel.MemberOverload(
+				"version_is_less_than",
+				[]*cel.Type{cel.StringType, cel.StringType},
+				cel.BoolType,
+				cel.FunctionBinding(versionIsLessThan)),
+		),
+	}
+}
+```
+
+This `ManagedClusterLib` will be added when `cel.NewEnv()`.
+
+```go
+func NewEvaluator() (*Evaluator, error) {
+	env, err := cel.NewEnv(
+		cel.Lib(ManagedClusterLib{}), // Add the ManagedClusterLib to the CEL environment
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CEL environment: %w", err)
+	}
+	return &Evaluator{env: env}, nil
+}
+```
+
+### Examples
+
+1. The user can select clusters by Kubernetes version listed in `managedCluster.Status.version.kubernetes`.
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1beta1
+kind: Placement
+metadata:
+  name: placement1
+  namespace: default
+spec:
+  numberOfClusters: 3
+  clusterSets:
+    - prod
+  predicates:
+    - requiredClusterSelector:
+        celSelector:
+          celExpressions:
+            - managedCluster.Status.version.kubernetes == "v1.30.0"
+```
+2. The user can use CEL [Standard macros](https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros) and [Standard functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#standard-definitions) on the `managedCluster` fields. For example, select clusters with version 1.30.* and 1.31.* by regex..
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1beta1
+kind: Placement
+metadata:
+  name: placement1
+  namespace: default
+spec:
+  numberOfClusters: 3
+  clusterSets:
+    - prod
+  predicates:
+    - requiredClusterSelector:
+        celSelector:
+          celExpressions:
+            - managedCluster.metadata.labels["version"].matches('^1\\.(30|31)\\.\\d+$')
+```
+
+3. The user can use CEL customized functions `versionIsGreaterThan` and `versionIsLessThan` to select clusters by version comparison. For example, select clusters whose version > 1.13.0.
+
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1beta1
+kind: Placement
+metadata:
+  name: placement1
+  namespace: default
+spec:
+  numberOfClusters: 3
+  clusterSets:
+    - prod
+  predicates:
+    - requiredClusterSelector:
+        celSelector:
+          celExpressions:
+            - managedCluster.metadata.labels["version"].versionIsGreaterThan("1.30.0")
+```
+
+
+4. The user can use CEL customized function `score` to select clusters by `AddonPlacementScore`. For example, select clusters whose cpuAvailable score < 2.
+
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1beta1
+kind: Placement
+metadata:
+  name: placement1
+  namespace: default
+spec:
+  numberOfClusters: 3
+  clusterSets:
+    - prod
+  predicates:
+    - requiredClusterSelector:
+        celSelector:
+          celExpressions:
+            - managedCluster.score("default, "cpuAvailable") < 2
```

Comment:
could we do like managedCluster.score.default.cpuAvailable? or we have to use a function here?

Reply:
managedCluster does not have a score field so if we pass `managedCluster.score.default.cpuAvailable` to the cel for evaluation, it can not recognize it.

Maybe we can use something like managedCluster.score(default).cpuAvailable. function score("crname") will return a map of [scorename]scorevalue, and user can use standard cel expression to select it.

Reply:
what if we do clusterScore instead of managedCluster instead? will it work?

Reply:
Hmm, have considered about this, when evaluate a expression is true or false, the clusterScore may still need a cluster name as input, managedCluster.score("default") equals to score(managedCluster, "default") (but can not write a cel expression like this).